Pass correct content argument to enter transforms

jsnajdr opened the pull request:
Fixed a bug that I found when testing the Table of Contents block (namely its refactoring in #54094). Steps to reproduce:
1. Create a Table of Contents block.
2. Type `/h2`<kbd>Enter</kbd> in the slash inserter to create a new Heading block.
3. Note that the Table of Contents block crashes when trying to add the new Heading to the TOC.

<img width="471" alt="Screenshot 2023-09-01 at 13 51 45" src="https://github.com/WordPress/gutenberg/assets/664258/4b77d02c-c653-4e79-9907-42f7b559e3ed">

This happens because the Heading block's `enter` transform creates the `core/heading` block with an incorrect `content` attribute. It should be a string, but instead it's an object, `{ content: 'foo' }`.

The fix is to:
1. Call the `transform` function with a `content` argument that is a string, not with an `attributes` argument that is a `{ content: string }` object. That's how the `enter` transform [is documented](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-transforms.md#enter), and also how the similar `prefix` transform is documented and implemented.
2. Ignore the `content` parameter when creating the `core/heading` block. The `/h2` value is not interesting and we don't want it to become the heading's text. Similar transforms for `core/code` and `core/separator` also ignore the `content` parameter.

There might be backward compatibility issues with this fix. If a plugin registers an `enter` transform that uses the `content` parameter, and assumes it to be an `attributes`-like object, it will stop working. Maybe, for compatibility, we should create a special class that acts both as a string and as an object with a `.content` attribute:
```js
class CString extends String {
  get content() { return this; }
}
transform( new CString( value.text ) );
```
Then both `transform( content ) {}` and `transform( { content } ) {}` is possible.

jsnajdr committed:
Pass correct content argument to enter transforms

Mamaduka reviewed (commented):
I agree that this would be a breaking change. The code has been in the core for years now.

Maybe we should update the documentation; note that callback receives an `attributes`-like object, plus keep the heading transformation fix.

Cc-ing @ellatrix and @ntsekouras as they are more familiar with the code in question.

jsnajdr committed:
Revert the change to public API

jsnajdr commented:
> I agree that this would be a breaking change. The code has been in the core for years now.

Yes, that's true, I found on wpdirectory.net that the TablePress plugin is actively using the current API:

<img width="591" alt="Screenshot 2023-09-01 at 15 02 05" src="https://github.com/WordPress/gutenberg/assets/664258/b66ddd7f-bb5e-4ee8-a484-1732c207d8cb">

I reverted the change that modifies the `transform` API, and instead I changed the docs to describe what really happens.

Mamaduka reviewed (approved):
Thank you, @jsnajdr ✅

This fixed the bug, and no breaking changes. The docs update also good to me.
